external/fv3fit/fv3fit/_shared/pytorch/graph.py

```diff
@@ -0,0 +1,205 @@
+import torch
+import tensorflow as tf
+import numpy as np
+import dataclasses
+from Building_Graph import BuildingGraph
+from hyperparameters import Hyperparameters
+from toolz.functoolz import curry
+from fv3fit._shared.pytorch.graphPredict import PytorchModel
+from fv3fit._shared.pytorch.Building_Graph import graphStruc
+from fv3fit._shared.pytorch.graph_configUnet import graphnetwork, GraphNetworkConfig
+from fv3fit._shared.pytorch.Graphloss import LossConfig
+from fv3fit._shared.pytorch.GraphOptim import OptimizerConfig
+from fv3fit._shared.pytorch.training_loop import TrainingLoopConfig
+
+# from ..tfdataset import iterable_to_tfdataset
+
+from fv3fit._shared import register_training_function
+from typing import (
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Mapping,
+)
+from fv3fit.tfdataset import select_keys, ensure_nd, apply_to_mapping
+
+
+@dataclasses.dataclass
+class GraphHyperparameters(Hyperparameters):
+    """
+    Args:
+        input_variables: names of variables to use as inputs
+        output_variables: names of variables to use as outputs
+        optimizer_config: selection of algorithm to be used in gradient descent
+        graph_network: configuration of graph network
+        training_loop: configuration of training loop
+        loss: configuration of loss functions, will be applied separately to
+            each output variable
+    """
+
+    input_variables: List[str]
+    output_variables: List[str]
+    optimizer_config: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("AdamW")
+    )
+    build_graph: BuildingGraph = dataclasses.field(
+        default_factory=lambda: BuildingGraph()
+    )
+
+    graph_network: GraphNetworkConfig = dataclasses.field(
+        default_factory=lambda: GraphNetworkConfig()
+    )
+    training_loop: TrainingLoopConfig = dataclasses.field(
+        default_factory=lambda: TrainingLoopConfig()
+    )
+    loss: LossConfig = LossConfig(loss_type="mse")
+
+    @property
+    def variables(self) -> Set[str]:
+        return set(self.input_variables).union(self.output_variables)
+
+
+# Some temporary transforms to make life easy for training process
+@curry
+def select_keys_mapping(
+    variable_names: Sequence[str], data: Mapping[str, tf.Tensor]
+) -> Mapping[str, tf.Tensor]:
+    return {name: data[name] for name in variable_names}
+
+
+@curry
+def normalize(
+    means: Mapping[str, float], stds: Mapping[str, float], data: Mapping[str, tf.Tensor]
+) -> Mapping[str, tf.Tensor]:
+    normed_data = {}
+    for key, tensor in data.items():
+        normed_data[key] = (tensor - means[key]) / stds[key]
```

Comment:
We found it was important in normalizing to add an `epsilon` value to standard deviations in the case of very small values (e.g. top of atmosphere moisture is almost always zero). Could you make sure the normalization you use here matches the normalization done by e.g. the dense model training?